### PR TITLE
feat: Add option to skip version bumping.

### PR DESCRIPTION
**Pull request by tomfrenken**
Hey all,

in our pipeline we already bumped the version of our package.json in a previous step, therefore we would like to skip bumping the version in this action.

This PR should introduce a simple option to do so.

### CODE REVIEW

**Comment by tomfrenken** [comment]
Hey @TriPSs, do you think we could add this option? :)

**Comment by TriPSs** [comment]
How different would this be than setting version file to an empty string?

**Comment by tomfrenken** [comment]
Well, with an empty string we would cause the following error:
```js
throw new Error(`File extension "${ext}" from file "${file}" is not supported`)
```
And even if we use `skip-version-file` and instead set a fallback-version to the one we actually want to use, the version is still bumped as is.

```js
if (skipVersionFile || skipCommit) {
      core.info('Using GIT to determine the new version')
      const versioning = await handleVersioningByExtension( // this function bumps the version currently
        'git',
        versionFile,
        versionPath,
        recommendation.releaseType,
      )
```

**Comment by tomfrenken** [comment]
I just added another conditional to set the `newVersion` to the `oldVersion`, as the `newVersion` property was only set when you are bumping (which we are trying not to do), and then later on this property is used for the changelog generation and therefore necessary.

File: src/index.js [hunk 1]
```diff
@@ -120,11 +127,12 @@ async function run() {
...
-        recommendation.releaseType
+        recommendation.releaseType,
+        skipBump
       )
 
-      newVersion = versioning.newVersion
       oldVersion = versioning.oldVersion
+      newVersion = skipBump ? oldVersion : versioning.newVersion
 
     } else {
```
Comment (TriPSs) [inline src/index.js:136]:
```suggestion
    } else if (!skipBump) {
```

**Review by TriPSs — CHANGES_REQUESTED** [review]
Although I find it a bit weird I do not see any harm in support things, could you do the following:
1. Commit to this repo following conventional commits
2. Process comments
3. Add tests

File: src/index.js (hunk 1 above)
Comment (TriPSs) [reply src/index.js:136]: Think this is cleaner than passing it down to an other function that than also does not do anything.

Comment (TriPSs) [reply src/index.js:136]: Ignore this, you need it for the old version, sorry.

**Comment by tomfrenken** [comment]
Thanks for the review, it wasn't entirely clear to me how all the testing in this repo behaves, but I believe my implementation should test it properly according to the existing standards.

Regarding your first point:  
> Commit to this repo following conventional commits

I did create the PR with a conventional commit message and would usually squash merge & remove all commit messages while squashing, would that work for you, or how exactly should I proceed?

**Comment by tomfrenken** [comment]
Also please approve the workflow run again :)

**Comment by TriPSs** [comment]
> I did create the PR with a conventional commit message and would usually squash merge & remove all commit messages while squashing, would that work for you, or how exactly should I proceed?

We could do that, regarding the test: should the tag it creates not be the same as the version in the file?

**Comment by TriPSs** [comment]
The test is also failing, it's creating a tag with `vnull` now

**Comment by tomfrenken** [comment]
Indeed, both the test and the implementation need some changes, currently the old version is only loaded after using the bump function, I'll make some more changes.

**Comment by tomfrenken** [comment]
Alright, I've adjusted the test as per your suggestion and also changed the classes to use a proper constructor instead of the `init` method, now the files are read and parsed at initialization, this way the `oldVersion` is accessible independently of bumping.

**Comment by TriPSs** [comment]
@tomfrenken you can also run the actions locally, see the [README](https://github.com/TriPSs/conventional-changelog-action?tab=readme-ov-file#development). My preference would be to only implement your feature here and leave existing code more as is (So keep the init, as I prefer to only pass args once and not for each class separate).

**Comment by tomfrenken** [comment]
Alright, thanks for the hint, however, the logs (when running the tests locally in parallel) are somewhat annoying to read 😬.

I've run the action now on my fork, all tests are passing.
@TriPSs as per your suggestion I removed the constructor again and added the init back, however with a slight adjustment as the init looks slightly different for the different file types.

**Review by TriPSs — COMMENTED** [review]
We could also make a `readFile` method in the base that is abstract, and call it inside the `init` of `BaseVersioning`, that way we do not need to define an init in each version.

What you think?

**Comment by tomfrenken** [comment]
@TriPSs thanks for the abstract method suggestion, great idea.
I've implemented the change and also added additional tests for corner cases.

File: src/index.js [hunk 2]
```diff
@@ -120,11 +124,17 @@ async function run() {
...
-        recommendation.releaseType
+        recommendation.releaseType,
+        skipBump
       )
 
-      newVersion = versioning.newVersion
       oldVersion = versioning.oldVersion
+      // If we are skipping the bump, we either use the fallback version or the old version as the new version.
+      if(skipBump){
+        newVersion = fallbackVersion && !oldVersion ? fallbackVersion : oldVersion
```
Comment (TriPSs) [inline src/index.js]:
```suggestion
        newVersion = oldVersion || fallbackVersion 
```

Comment (TriPSs) [reply src/index.js]: We could also do
```suggestion
        newVersion = skipBump
                ? oldVersion || fallbackVersion 
                : versioning.newVersion
```

File: src/index.js [hunk 3]
```diff
@@ -135,12 +145,16 @@ async function run() {
...
       )
-
-      newVersion = versioning[0].newVersion
       oldVersion = versioning[0].oldVersion
+      // If we are skipping the bump, we either use the fallback version or the old version as the new version.
+      if(skipBump){
+        newVersion = fallbackVersion && !oldVersion ? fallbackVersion : oldVersion
+      } else {
+        newVersion = versioning[0].newVersion
+      }
```
Comment (TriPSs) [inline src/index.js]: Is this needed as `handleVersioningByExtension` makes sure the correct version is set in `newVersion`?

File: src/version/json.js [hunk 4]
```diff
@@ -6,33 +6,38 @@ const bumpVersion = require('../helpers/bumpVersion')
 
 module.exports = class Json extends BaseVersioning {
 
+  eol = null;
+  jsonContent = {};
```
Comment (TriPSs) [inline src/version/json.js]:
```suggestion
  eol = null
  jsonContent = {}
```

**Review by TriPSs — COMMENTED** [review]
Thanks already for your work!

File: src/index.js (hunk 3 above)
Comment (tomfrenken) [reply src/index.js]: Yeah, that's a good point, I think I should move the entire versioning logic into that function instead of putting it here

**Comment by tomfrenken** [comment]
Awesome, moving the logic to the version function also removed the code duplication, and the improved conditional logic is also way cleaner.

**Review by TriPSs — APPROVED** [review]

**Comment by TriPSs** [comment]
Thanks for the PR 🎉

**Comment by tomfrenken** [comment]
Thanks for the quick reviews!
